ELPP-2252 KVS Cache for transformer

stephenwf opened the pull request:
@giorgiosironi  @thewilkybarkid

giorgiosironi commented:
I understand the need of storing subjects data, but given a single SQS queue how do we avoid doing this for all types that come out from its messages? Is there some routing done from the client code of the library? Will reread it tomorrow morning.

stephenwf commented:
yes so here in search:
https://github.com/elifesciences/search/blob/develop/src/Search/Gearman/Command/QueueWatchCommand.php#L49-L50

We have 2 transforms, one for `QueueItem` to gearman task and another `QueueItem` to ApiSDK item. We can add to those lines: 

```php
if($item->getType() === 'subject') {
    $entity = $this->cachedTransformer->transform($item);
}
```

Or something along those lines.

Alternatively I could add a callable "shouldCache" to the constructor that will dictate when to cache.

stephenwf commented:
Okay I've added the `$shouldCache` property, I think that will work

thewilkybarkid commented on src/Queue/CachedTransformer.php:
```diff
@@ -15,18 +15,37 @@ class CachedTransformer implements QueueItemTransformer
+
+    private function shouldCacheEntity(string $type, string $id)
+    {
+        if ($shouldCache = $this->shouldCache) {
```
The `if` is redundant isn't it?

stephenwf replied:
Ah forgot to make the should cache optional in the constructor, avoids something like `function() { return true; }` in recommendations

thewilkybarkid commented on src/Queue/CachedTransformer.php:
```diff
@@ -15,18 +15,37 @@ class CachedTransformer implements QueueItemTransformer
 
     public function get($id, $type)
     {
+        if (!$this->shouldCacheEntity($id, $type)) {
```
This should probably come last (what if the cache entry has disappeared?).

giorgiosironi commented:
I still do not understand the integration with the client project. Using a cache inside `queue:watch` does not seem a good idea because `queue:watch` is notified of changes, hence if it then caches the GET requests it performs a change can be lost due to a stale cache?

I thought this kind of cache should be written and read by the HTTP processes that returns responses to the API calls (to fill in `subjects`-related fields in the response), not by the long-running processes?

stephenwf commented:
@giorgiosironi  So the `queue:watch` will only use the `transform()` command like it does just now, when it calls that it will always 100% make a network request to get the item. If its been told to cache that item it will add it to the cache. 

Only the user-facing controller will use the `get()` method hitting the cache

giorgiosironi commented:
But why tying up this cache with the Transformers and the queue at all, if we only need it on the serving side (which I assume does not use Transformers and these classes)? It can save some dozen calls to populate the cache from the transform() that can then be used in the get(), but it's a lot of complexity for little gain?

stephenwf commented:
For recommendations we need a cache that will server 99% of the content imported, the calls to populate can be handled in the cache itself, for the initial import we can have cache that will batch the calls so its not trashing the kvs. The gain is simply cache invalidation, with this we can always ensure fresh content. (without Dos-ing journal 👀 )

giorgiosironi commented on src/Queue/CachedTransformer.php:
```diff
@@ -0,0 +1,108 @@
+use eLife\ApiSdk\ApiSdk;
+use eLife\ApiSdk\Model\Model;
+use Psr\Log\LoggerInterface;
+
```
could you add a docblock here that explains the difference between `get()` and `refresh()` and the use cases from which they are called?

giorgiosironi replied:
want this to be clear to people reading it in months time

giorgiosironi commented on src/Queue/CachedTransformer.php:
```diff
@@ -0,0 +1,108 @@
+        return true;
+    }
+
+    public function get($id, $type)
```
an interface for this `get()` method would be useful as code in the serving phase of the service will not depend on a `CachedTransformer` but rather on a smaller interface

stephenwf commented:
@giorgiosironi I've added an interface and changed up the classes that implement the interface. Its possible these two bits of functionality could be split out if that makes more sense in each implementation. 

Also described the 3 scenarios for the `get()` command, cache, no-cache and cache but stale.

giorgiosironi commented on src/Queue/CachedTransformer.php:
```diff
@@ -0,0 +1,131 @@
+
+    public static function getKey(string $type, string $id) : string
+    {
+        return sha1(sha1($id), sha1($type));
```
why not just `sha1("$type}:{$id}")` :thinking:

thewilkybarkid replied:
Second argument of `sha1()` is a boolean too.

stephenwf replied:
👍  sorry, fixed

giorgiosironi commented:
Segregated interfaces do not necessarily imply segregated implementations, good to go :+1:
